AJ-1309: Allow GCP aggregated workspace

jladieu opened the pull request:
Ticket: [AJ-1309](https://broadworkbench.atlassian.net/browse/AJ-1309)

This PR includes three refactoring commits which should lower the barrier to using `AggregatedWorkspaceService` in `SnapshotService`.

---

**PR checklist**

- [x] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala). _not doing this as it's an internal only change_
- [x] Get two thumbsworth of PR review
- [x] Verify all tests go green, including CI tests
- [x] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [x] Inform other teams of any substantial changes via Slack and/or email _(n/a, this is a minor refactoring change that unlocks more functional changes)_


[AJ-1309]: https://broadworkbench.atlassian.net/browse/AJ-1309?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jladieu committed:
AJ-1309: Honor the Rawls `WorkspaceType` when creating a WSM workspace just in time as part of creating a Snapshot.

The existing logic blindly creates a workspace without consulting the Rawls `WorkspaceType`, so it will defaulted to having a Rawls stage in WSM even if what is _in_ Rawls is an MC workspace.

Honoring the `WorkspaceType` should help avoid having the data in WSM drift from what is in Rawls going forward as this logic evolves to eventually do a better job supporting Azure.

This is in preparation for making changes to `AggregatedWorkspaceService`, which will start to be used by `SnapshotService` for both GCP and Azure workspaces.

jladieu committed:
AJ-1309: Add support for `GCP` MC workspaces to `AggregatedWorkspaceService`.

In preparation for improving the validation logic in `SnapshotService`, this unblocks modifying `AggregatedWorkspaceService` to be used with GCP workspaces from WSM as well, to determine `CloudPlatform` and whether or not a workspace needs to be created just in time.

jladieu commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -759,6 +757,8 @@ case class AzureManagedAppCoordinates(tenantId: UUID,
                                       landingZoneId: Option[UUID] = None
 )
 
+case class GcpProjectInfo(projectId: GoogleProjectId)
```
Does this merit publishing a new `rawls-model`?

aherbst-broad replied:
I think this is only being consumed internally by the AggregatedWorkspaceService, so I'd say it's not necessary.

jladieu commented:
I'd welcome advice on whether and what kind of verification would be worthwhile above and beyond the unit tests.  The changed code is used by `ListWorkspaces` and `GetWorkspace` so I'm assuming poking around the UI in Workspaces is a good place to start.

jladieu commented:
After discussing with @davidangb and realizing that, while by design `WorkspaceType` and `Azure/GCP` are orthogonal, in practice `WorkspaceType.RAWLS` === `GCP` and `WorkspaceType.MCWorkspace` === `Azure` -- so these changes amount to futureproofing only.  That said, it still seemed like a net positive, so I'm rolling ahead with these.

jladieu commented on core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:
```diff
@@ -72,7 +72,7 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
   // Ideally this would rely on Terra Policy Service, but until TPS is enabled for GCP
   // We'll have to use this workaround for identifying protected status
-  def validateProtectedStatus(workspaceContext: Workspace, snapshot: NamedDataRepoSnapshot): Unit =
+  private def validateProtectedStatus(workspaceContext: Workspace, snapshot: NamedDataRepoSnapshot): Unit =
```
Just opportunistic warning tidying.

jladieu commented on core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:
```diff
@@ -108,7 +108,7 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
 
   // AS-787 - rework the data so that it's in the same place in the JSON with a list and get snapshot responses
-  def massageSnapshots(references: ResourceList): SnapshotListResponse = {
+  private def massageSnapshots(references: ResourceList): SnapshotListResponse = {
```
Just opportunistic warning tidying.

jladieu commented on core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:
```diff
@@ -127,7 +127,7 @@ class SnapshotService(protected val ctx: RawlsRequestContext,
       // to having no references, so return the empty list.
       case Failure(ex: bio.terra.workspace.client.ApiException) if ex.getCode == 404 =>
-        new SnapshotListResponse(Seq.empty[DataRepoSnapshotResource])
+        SnapshotListResponse(Seq.empty[DataRepoSnapshotResource])
```
Just opportunistic warning tidying.

aherbst-broad commented on core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceService.scala:
```diff
@@ -38,46 +40,42 @@ class AggregatedWorkspaceService(workspaceManagerDAO: WorkspaceManagerDAO) exten
+                convertPolicies(wsmInfo)
               )
-            case None =>
+            case (Some(gcpContext), None) =>
```
👍 this should put us in a better place for real MC GCP workspaces.

aherbst-broad reviewed (approved):
This generally makes sense, thanks for doing this.

jladieu committed:
Refactor `AggregatedWorkspaceService`

* Rename `getAggregatedWorkspace` to `optimizedFetchAggregatedWorkspace`
to make explicit that the WSM call is being skipped for Rawls
workspaces.
* This becomes an overload of `fetchAggregatedWorkspace`, which is
extracted to bypass the optimization and provide a way to fetch WSM info
for a Rawls workspace.  This will be used in a subsequent PR by
`SnapshotService`.

jladieu commented:
@aherbst-broad @davidangb I added a third commit to this PR and reset the reviews; this third commit provides the access path for `SnapshotService` to fetch a Rawls based workspace using the 'unoptimized' version of what used to be named `getAggregatedWorkspace` and is now `optimizedFetchAggregatedWorkspace`.

davidangb commented on core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/HttpWorkspaceManagerDAO.scala:
```diff
@@ -50,8 +56,16 @@ class HttpWorkspaceManagerDAO(apiClientProvider: WorkspaceManagerApiClientProvid
+      case WorkspaceType.RawlsWorkspace => WorkspaceStageModel.RAWLS_WORKSPACE
+      case WorkspaceType.McWorkspace    => WorkspaceStageModel.MC_WORKSPACE
+    }
+    getWorkspaceApi(ctx).createWorkspace(new CreateWorkspaceRequestBody().id(workspaceId).stage(stage))
```
My first reaction: how did this ever work for MC_WORKSPACEs previously????

Looking deeper at it, `MultiCloudWorkspaceService` never calls this `createWorkspace`; it calls `createWorkspaceWithSpendProfile` instead. I wonder if these two methods are worth combining. That's out of scope for this PR (we could yak-shave forever).

jladieu replied:
Agreed to punt this....

davidangb commented on model/src/main/scala/org/broadinstitute/dsde/rawls/model/WorkspaceModel.scala:
```diff
@@ -759,6 +757,8 @@ case class AzureManagedAppCoordinates(tenantId: UUID,
                                       landingZoneId: Option[UUID] = None
 )
 
+case class GcpProjectInfo(projectId: GoogleProjectId)
```
`GoogleProjectId` is already a wrapper around `String` for type safety - do we actually need this additional layer of wrapper?

If we do need it, I note that we don't have a `JsonFormat` for `GcpProjectInfo`, so we would likely fail to de/serialize it in JSON requests/responses. I don't think `GcpProjectInfo` is used anywhere that needs serdes, so this is ok, just raising it to keep an eye on

jladieu replied:
I'd love to avoid touching this file, although I thought maybe we'd eventually (YAGNI?) cram more stuff into `GcpProjectInfo`...is this potentially a landmine for the next person if I leave it in here without `JsonFormat`?

jladieu replied:
Ok dropped it...very happy with the outcome!  Thanks for the suggestion.

davidangb reviewed (approved):
:+1: and thank you for this. Two comments inline, neither of which should block merging.

jladieu commented:
/jenkins retest
